### DISCUSSION TITLE
A question on the dim and dtype of action.

### DISCUSSION THREAD

**Discussion by lk1983823** [johbrust/simulink_gym#7] [Q&A] [answered]
I have successfully run the cartpole_simulink.ipynb file. Now I have two questions:
1. In the step method in cartpole_simulink.py, I find you define action = int(action) in line 119. Is it due to just the action.space requirentment or the TCP/IP sending requirements. If I train a RL model with float type action, will it have any problem?
2. In the cartpole example, the dimension of action is only one. If I have action with multiple dimensions, will there be any fix in simulink_block_lib? Can you help to specify how to fix it? Or suggest to add an example with more than one dimension with float dtype. 
Thanks.

**Comment by johbrust** [comment, marked as answer]
Regarding question 1: Yes, `action = int(action)` is specific to the environment and its action space. Floats are no problem.

Regarding question 2: You can define an input vector (array with one dimension) with multiple actions. But you have to set this also in your Simulink model in the TCP/IP Input block (see description [here](https://github.com/johbrust/simulink_gym/tree/main/simulink_block_lib#tcpip-in-block) for block parameter *Number of input signals*).

I created an example [here in a separate branch](https://github.com/johbrust/simulink_gym/tree/examples/simple_reacher_simulink/examples/envs/simple_reacher_simulink) with 2 float actions for you to check out.

**Reply by lk1983823** [reply]
> Regarding question 1: Yes, `action = int(action)` is specific to the environment and its action space. Floats are no problem.
> 
> Regarding question 2: You can define an input vector (array with one dimension) with multiple actions. But you have to set this also in your Simulink model in the TCP/IP Input block (see description [here](https://github.com/johbrust/simulink_gym/tree/main/simulink_block_lib#tcpip-in-block) for block parameter _Number of input signals_).
> 
> I created an example [here in a separate branch](https://github.com/johbrust/simulink_gym/tree/examples/simple_reacher_simulink/examples/envs/simple_reacher_simulink) with 2 float actions for you to check out.

Thanks ! The toy example is very helpful !

**Comment by lk1983823** [comment]
I add step_size in this simple_reacher_simulink in the attachment and run it successfully.
[simple_reacher_simulink_with_step_size.tar.gz](https://github.com/johbrust/simulink_gym/files/10131973/simple_reacher_simulink_with_step_size.tar.gz)

**Reply by johbrust** [reply]
The difference is that `Simulink.Parameter` allows for more detailed specifications, as you can see below. You can specify minimum and maximum values or a unit and give a description of the parameter.

![grafik](https://user-images.githubusercontent.com/16197185/205038713-96c71839-ab2c-4367-8c17-091988911849.png)

You can create new a `Simulink.Parameter` from the block parameter dialogue by selecting `Simulink.Parameter` for the *Value* in the *Create* dialogue:

![grafik](https://user-images.githubusercontent.com/16197185/205039229-5303755a-2fc3-48d5-a2fa-9713a3e65477.png)

You can also create parameters in the model explorer through *Add* &rarr; *Simulink Parameter* and afterwards select it in the block dialogue:

![grafik](https://user-images.githubusercontent.com/16197185/205039398-e8b178d9-300d-4cd2-b96a-63bcdb40a847.png)

**Reply by lk1983823** [reply]
> The difference is that `Simulink.Parameter` allows for more detailed specifications, as you can see below. You can specify minimum and maximum values or a unit and give a description of the parameter.
> 
> ![grafik](https://user-images.githubusercontent.com/16197185/205038713-96c71839-ab2c-4367-8c17-091988911849.png)
> 
> You can create new a `Simulink.Parameter` from the block parameter dialogue by selecting `Simulink.Parameter` for the _Value_ in the _Create_ dialogue:
> 
> ![grafik](https://user-images.githubusercontent.com/16197185/205039229-5303755a-2fc3-48d5-a2fa-9713a3e65477.png)
> 
> You can also create parameters in the model explorer through _Add_ → _Simulink Parameter_ and afterwards select it in the block dialogue:
> 
> ![grafik](https://user-images.githubusercontent.com/16197185/205039398-e8b178d9-300d-4cd2-b96a-63bcdb40a847.png)

Thanks! I have updated the attachment.
